lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -2242,9 +2242,47 @@ struct IntelLSUControlsInfo {
   llvm::Optional<unsigned> PrefetchInfo;
 };
 
-IntelFPGADecorations
-tryParseIntelFPGAAnnotationString(StringRef AnnotatedCode) {
-  IntelFPGADecorations Decorates;
+void processOptionalAnnotationInfo(Constant *Const,
+                                   std::string &AnnotationString) {
+  // Handle optional parameter. It can be for example
+  // { %struct.S, i8*, void ()* } { %struct.S undef, i8* null,
+  //                                void ()* @_Z4blahv }
+  // Now we will just handle integer constants (wrapped in a constant
+  // struct, that is being bitcasted to i8*), converting them to string.
+  // TODO: remove this workaround when/if an extension spec that allows or adds
+  // variadic-arguments UserSemantic decoration
+  if (!Const->getNumOperands())
+    return;
+  if (auto *CStruct = dyn_cast_or_null<ConstantStruct>(Const->getOperand(0))) {
+    uint32_t NumOperands = CStruct->getNumOperands();
+    if (!NumOperands)
+      return;
+    if (auto *CInt = dyn_cast<ConstantInt>(CStruct->getOperand(0))) {
+      AnnotationString += ": ";
+      AnnotationString += std::to_string(CInt->getSExtValue());
+    }
+    for (uint32_t I = 1; I != CStruct->getNumOperands(); ++I) {
+      if (auto *CInt = dyn_cast<ConstantInt>(CStruct->getOperand(I))) {
+        AnnotationString += ", ";
+        AnnotationString += std::to_string(CInt->getSExtValue());
+      }
+    }
+  }
+}
+
+AnnotationDecorations tryParseAnnotationString(SPIRVModule *BM,
+                                               StringRef AnnotatedCode) {
+  AnnotationDecorations Decorates;
+  const bool AllowFPGAMemAccesses =
+      BM->isAllowedToUseExtension(ExtensionID::SPV_INTEL_fpga_memory_accesses);
+  const bool AllowFPGAMemAttr = BM->isAllowedToUseExtension(
+      ExtensionID::SPV_INTEL_fpga_memory_attributes);
+  if (!(AllowFPGAMemAccesses || AllowFPGAMemAttr)) {
```

Comment:
It is totally up to you, but reverse condition kinda looks easier to understand
```suggestion
  if (!AllowFPGAMemAccesses && !AllowFPGAMemAttr) {
```